datasets/flwr_datasets/partitioner/grouped_natural_id_partitioner.py

```diff
@@ -0,0 +1,213 @@
+        - "allow-bigger": The first group can be bigger than the group_size.
+        - "drop-reminder": The last group will be dropped if it is smaller than the
+        group_size.
+        - "strict": Raises a ValueError if the remainder is not zero. In this mode you
+        expect each group to have the same size.
+    sort_unique_ids: bool
+        If True, the unique natural ids will be sorted before creating the groups.
+
+    Examples
+    --------
+    "sentiment140" (aka Twitter) dataset
+
+    >>> from flwr_datasets import FederatedDataset
+    >>> from flwr_datasets.partitioner import GroupedNaturalIdPartitioner
+    >>>
+    >>> partitioner = GroupedNaturalIdPartitioner(partition_by="user", group_size=2)
+    >>> fds = FederatedDataset(dataset="sentiment140",
+    >>>                        partitioners={"train": partitioner})
+    >>> partition = fds.load_partition(0)
+    """
+
+    def __init__(
+        self,
+        partition_by: str,
+        group_size: int,
+        mode: Literal[
+            "allow-smaller", "allow-bigger", "drop-reminder", "strict"
+        ] = "allow-bigger",
+        sort_unique_ids: bool = False,
+    ) -> None:
+        super().__init__()
+        self._partition_id_to_natural_ids: Dict[int, List[Any]] = {}
+        self._natural_id_to_partition_id: Dict[Any, int] = {}
+        self._partition_id_to_indices: Dict[int, NDArrayInt] = {}
+        self._partition_by = partition_by
+        self._mode = mode
+        self._sort_unique_ids = sort_unique_ids
+
+        if group_size < 0:
+            raise ValueError("group_size must be a positive integer")
+        self._group_size = group_size
+
+    def _create_int_partition_id_to_natural_id(self) -> None:
+        """Create a mapping from int indices to unique client ids from dataset.
+
+        Natural ids come from the column specified in `partition_by`.
+        """
+        unique_natural_ids = self.dataset.unique(self._partition_by)
+        if self._sort_unique_ids:
+            unique_natural_ids = sorted(unique_natural_ids)
+        num_unique_natural_ids = len(unique_natural_ids)
+        remainder = num_unique_natural_ids % self._group_size
+
+        if self._mode == "allow-bigger":
+            num_groups = num_unique_natural_ids // self._group_size
+            groups_of_natural_ids = np.array_split(unique_natural_ids, num_groups)
+        elif self._mode == "drop-reminder":
+            num_groups = num_unique_natural_ids // self._group_size
+            # Narrow down the unique_natural_ids to not have a bigger group
+            # which is the behavior of the np.array_split
+            unique_natural_ids = unique_natural_ids[
+                : int(num_groups * self._group_size)
+            ]
+            groups_of_natural_ids = np.array_split(unique_natural_ids, num_groups)
+        elif self._mode == "allow-smaller":
+            num_groups = num_unique_natural_ids // self._group_size
```

Comment:
minor observation: it seems this computation is needed for all models -> should we move it to right under the `reminder` calculation?

Reply:
Yes, thanks. I'll change it.

Reply:
Done